### PR TITLE
Create Working with UICollectionView guide

### LINKED ISSUE #527: New guide for UICollectionView (follow-up to #409)
In light of #409, we should publish updates to the Guides, or maybe a new guide?

Proposal, new guide:

> ### Working with `UICollectionView`
>
> - details the history (legacy) of `IGListCollectionView`
> - links to #409 
> - lists the methods **you should never call** on `UICollectionView` when using ListKit

Thoughts?

### PR DESCRIPTION
**Pull request by jessesquires**
Close #527

### CODE REVIEW

File: Guides/Working with UICollectionView.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
+# Working with `UICollectionView`
+
+This guide provides details on how to work with [`UICollectionView`](https://developer.apple.com/reference/uikit/uicollectionview) and `IGListKit`.
+
+## Background
+
+Early versions of `IGListKit` (v2.x and prior) shipped with a subclass of `UICollectionView` called [`IGListCollectionView`](https://github.com/Instagram/IGListKit/blob/2.1.0/Source/IGListCollectionView.h). The class contained *no* special functionality and was merely used to enforce compile-time restrictions to prevent users from calling certain methods directly on `UICollectionView`. Beginning with v3.0, `IGListCollectionView` [was removed](https://github.com/Instagram/IGListKit/commit/2284ce389708f62d99f48ff2ec15644f1ec59537) for a number of reasons.
```
Comment (rnystrom) [inline Guides/Working with UICollectionView.md]: nit: s/v3.0/3.0

File: Guides/Working with UICollectionView.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
+# Working with `UICollectionView`
+
+This guide provides details on how to work with [`UICollectionView`](https://developer.apple.com/reference/uikit/uicollectionview) and `IGListKit`.
+
+## Background
+
+Early versions of `IGListKit` (v2.x and prior) shipped with a subclass of `UICollectionView` called [`IGListCollectionView`](https://github.com/Instagram/IGListKit/blob/2.1.0/Source/IGListCollectionView.h). The class contained *no* special functionality and was merely used to enforce compile-time restrictions to prevent users from calling certain methods directly on `UICollectionView`. Beginning with v3.0, `IGListCollectionView` [was removed](https://github.com/Instagram/IGListKit/commit/2284ce389708f62d99f48ff2ec15644f1ec59537) for a number of reasons. 
+For further discussion see [#240](https://github.com/Instagram/IGListKit/issues/240) and [#409](https://github.com/Instagram/IGListKit/issues/409).
```
Comment (rnystrom) [inline Guides/Working with UICollectionView.md]: newline?

File: Guides/Working with UICollectionView.md [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+## Methods to avoid
+
+One of the primary purposes of `IGListKit` is to perform optimal batch updates for `UICollectionView`. Thus, clients **should never** call any APIs on `UICollectionView` that involved reloading, inserting, deleting, or otherwise updating cells and index paths. Instead, use the APIs provided by [`IGListAdapter`](https://instagram.github.io/IGListKit/Classes/IGListAdapter.html). You should also avoid setting the [`delegate`](https://developer.apple.com/reference/uikit/uicollectionview/1618033-delegate) and [`dataSource`](https://developer.apple.com/reference/uikit/uicollectionview/1618091-datasource) of the collection view, as this is also the responsibility of `IGListAdapter`.
+
+Avoid calling the following methods:
+
+```objc
+// Objective-C
```
Comment (rnystrom) [inline Guides/Working with UICollectionView.md]: nit: remove IMO

File: Guides/Working with UICollectionView.md [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+- (void)setDelegate:(id<UICollectionViewDelegate>)delegate;
+
+- (void)setDataSource:(id<UICollectionViewDataSource>)dataSource;
+
+- (void)setBackgroundView:(UIView *)backgroundView;
+```
+
+## Performance
+
+In iOS 10, a new [cell prefetching API](https://developer.apple.com/reference/uikit/uicollectionviewdatasourceprefetching) was introduced. At Instagram, enabling this feature substantially degraded scrolling performance. We reccommend setting [`isPrefetchingEnabled`](https://developer.apple.com/reference/uikit/uicollectionview/1771771-isprefetchingenabled) to `NO` (`false` in Swift). Note that the default value is `true`.
```
Comment (rnystrom) [inline Guides/Working with UICollectionView.md]: "reccommend"

Maybe add a line about how IGListKit doesn't use this feature, so having it enabled is just passive waste

**Comment by Sherlouk** [comment]
Had a read and looks good to me!

I understand that prefetching had detrimental effect on Instagram (likewise on my apps we had to disable it to stop multiple issues on iOS 10) but we actually found that was because we didn't "support" it less that it's "not helpful".

Has anyone actually explored doing correct prefetching with IGListKit? Or maybe even adding a prefetching API into section controllers? (Might be a good discussion point)

**Pull request closed without merging**